hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -496,12 +504,79 @@ SqlResult execute(ExplainStatementPlan plan) {
+            // pruning doesn't get activated at all for this query.
+            if (!(strategy instanceof DefaultPartitioningStrategy)
+                    && !(strategy instanceof AttributePartitioningStrategy)) {
+                allVariantsValid = false;
+                break;
+            }
+
+            // ordering of attributes matters for partitioning (1,2) produces different partition than (2,1).
+            final List<String> orderedKeyAttributes = new ArrayList<>();
+            if (strategy instanceof AttributePartitioningStrategy) {
+                final var attributeStrategy = (AttributePartitioningStrategy) strategy;
+                orderedKeyAttributes.addAll(asList(attributeStrategy.getPartitioningAttributes()));
+            } else {
+                orderedKeyAttributes.add(DEFAULT_UNIQUE_KEY);
+            }
+
+            for (final Map<String, Expression<?>> perMapCandidate : perMapCandidates) {
+                Object[] partitionKey = new Object[orderedKeyAttributes.size()];
+                for (int i = 0; i < orderedKeyAttributes.size(); i++) {
+                    final String attribute = orderedKeyAttributes.get(i);
+                    if (!perMapCandidate.containsKey(attribute)) {
+                        // Shouldn't happen, defensive check in case Opt logic breaks and produces variants
+                        // that do not contain all the required partitioning attributes.
+                        throw new HazelcastException("Partition Pruning candidate"
+                                + " does not contain mandatory attribute: " + attribute);
+                    }
+
+                    partitionKey[i] = perMapCandidate.get(attribute).eval(null, evalContext);
+                }
+
+                // In case of default strategy there is only a single key, not an array.
+                // AttributePartitioningStrategy with single attribute also does not use array.
+                // TODO: this code must be in sync with logic in other places. Do this better.
+                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
+                        partitionKey.length == 1 ? partitionKey[0] : partitionKey);
+                if (partition == null) {
+                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
+                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
+                    allVariantsValid = false;
+                    break;
+                }
+                partitions.add(partition.getPartitionId());
+            }
+        }
+
+        if (!partitions.isEmpty() && allVariantsValid) {
+            // add partitionId required for allToOne("") routing to ClientSink
+            // TODO: this artificially adds redundant partition
+            partitions.add(hazelcastInstance.getPartitionService().getPartition("").getPartitionId());
```

Comment:
possible NPE if the cluster is mid-repartitioning/migration? This logic needs improvement anyway